v2/car.go

```diff
@@ -41,12 +46,90 @@ type (
 	}
 )
 
-// Size gets the size of Header in number of bytes.
-func (h *Header) Size() int {
-	return HeaderBytesSize
+// WriteTo writes this characteristics to the given writer.
+func (c Characteristics) WriteTo(w io.Writer) (n int64, err error) {
+	wn, err := writeUint64To(w, c.Hi)
+	if err != nil {
+		return
+	}
+	n += wn
+	wn, err = writeUint64To(w, c.Lo)
+	if err != nil {
+		return
+	}
+	n += wn
+	return
 }
 
 // Size gets the size of Characteristics in number of bytes.
-func (c *Characteristics) Size() int {
+func (c Characteristics) Size() uint64 {
 	return CharacteristicsBytesSize
 }
+
+// NewHeader instantiates a new CAR v2 header, given the byte length of a CAR v1.
+func NewHeader(carV1Size uint64) *Header {
+	header := &Header{
+		Characteristics: EmptyCharacteristics,
+		CarV1Size:       carV1Size,
+	}
+	header.CarV1Offset = PrefixBytesSize + HeaderBytesSize
+	header.IndexOffset = header.CarV1Offset + carV1Size
+	return header
+}
+
+// Size gets the size of Header in number of bytes.
+func (h *Header) Size() uint64 {
+	return HeaderBytesSize
+}
+
+// WithIndexPadding sets the index offset from the beginning of the file for this header and returns the
+// header for convenient chained calls.
+// The index offset is calculated as the sum of PrefixBytesLen, HeaderBytesLen,
+// Header#CarV1Len, and the given padding.
+func (h *Header) WithIndexPadding(padding Padding) *Header {
+	h.IndexOffset = h.IndexOffset + uint64(padding)
+	return h
+}
+
+// WithCarV1Padding sets the CAR v1 dump offset from the beginning of the file for this header and returns the
+// header for convenient chained calls.
+// The CAR v1 offset is calculated as the sum of PrefixBytesLen, HeaderBytesLen and the given padding.
+// The call to this function also shifts the Header#IndexOffset forward by the given padding.
+func (h *Header) WithCarV1Padding(padding Padding) *Header {
+	h.CarV1Offset = h.CarV1Offset + uint64(padding)
+	h.IndexOffset = h.IndexOffset + uint64(padding)
+	return h
+}
+
+// WriteTo serializes this header as bytes and writes them using the given io.Writer.
+func (h *Header) WriteTo(w io.Writer) (n int64, err error) {
+	wn, err := h.Characteristics.WriteTo(w)
+	if err != nil {
+		return
+	}
+	n += wn
+	wn, err = writeUint64To(w, h.CarV1Offset)
+	if err != nil {
+		return
+	}
+	n += wn
```

Comment:
nit: you only use `wn` when `err == nil`, and it's always `8`, so I think you could simplify these three uint64 writes to:

```
if err = writeUint64To(w, h.CarV1Offset); err != nil {
    return
}
n += 8
```

At which point `writeUint64To(w, x)` would just be shorthand for `binary.Write(w, binary.LittleEndian, x)`, so it's up to you if it's worth keeping to save a few keystrokes.

Reply:
The hard-coded 8 in the above approach is what I don't like, since going by encapsulation rules `writeUint64To` does not necessarily have to write 8 bits (even though it does that now) :) 

Having said that, I'll change that to what you suggested, since it is simpler and functionally equal

Reply:
If you think you'll write integers of different lengths, by all means do it some other way - I'm just working from what I see in the current code. Worth noting that I'm pretty sure cbor allows for shorter integers than 64-bit, so I wonder if the carv2 draft spec specifically says we should use 64 bits for all of these.